include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -0,0 +1,571 @@
+                          "The mate object must be a std::tuple of size 3 with "
+                          "1) a std::ranges::ForwardRange with a value_type modelling seqan3::alphabet_concept, "
+                          "2) an std::UnsignedIntegral, and"
+                          "3) an std::UnsignedIntegral.");
+
+        if constexpr (!detail::decays_to_ignore_v<mate_type>)
+            static_assert((std::ranges::ForwardRange<decltype(std::get<0>(mate))>            &&
+                          alphabet_concept<value_type_t<decltype(std::get<0>(mate))>>        &&
+                          std::UnsignedIntegral<remove_cvref_t<decltype(std::get<1>(mate))>> &&
+                          std::UnsignedIntegral<remove_cvref_t<decltype(std::get<2>(mate))>>),
+                          "The mate object must be a std::tuple of size 3 with "
+                          "1) a std::ranges::ForwardRange with a value_type modelling seqan3::alphabet_concept, "
+                          "2) an std::UnsignedIntegral, and"
+                          "3) an std::UnsignedIntegral.");
+
+        if constexpr (!detail::decays_to_ignore_v<tag_dict_type>)
+            static_assert(std::Same<remove_cvref_t<tag_dict_type>, sam_tag_dictionary>,
+                          "The tag_dict object must be of type seqan3::sam_tag_dictionary.");
+
+        // ---------------------------------------------------------------------
+        // logical Requirements
+        // ---------------------------------------------------------------------
+        static_assert(detail::decays_to_ignore_v<align_type> ||
+                      (!detail::decays_to_ignore_v<seq_type> && !detail::decays_to_ignore_v<offset_type>),
+                      "If you specify an align object you must also specify the seq and offset object.");
+
+        // ---------------------------------------------------------------------
+        // Writing
+        // ---------------------------------------------------------------------
+        if constexpr (!detail::decays_to_ignore_v<header_type>)
+        {
+            if constexpr(!detail::decays_to_ignore_v<ref_id_type>)
+            {
+                auto ref = header.ref_dict | view::get<0>
+                                           | ranges::view::remove_if([&ref_id] (std::string & s) {return s != ref_id;});
+                if (ref.begin() == ref.end()) // no reference id matched
+                    throw format_error(std::string("The ref_id '") + ref_id + "' was not in the list of references");
+            }
+
+            if(!written_header)
+            {
+                check_header_information(header);
+                write_header(stream, header, options);
+                written_header = true;
+            }
+
+            // if all other fields decay to std::ignore, only the header
+            // should be printed and not an all-empty-record.
+            if constexpr (detail::decays_to_ignore_v<seq_type> &&
```

Comment:
the header is only required in BAM format but optional in SAM format. We wanted to keep the header in the record because we don't want to disrupt the range-based functionality by the need of first calling a read/write header function, especially because for BAM files, reading the header is a must as correct parsing is not possible otherwise.
But when the user only calls file_out = {header} it would be very unintuitive if the header and an empty record is printed.. don't you think?